### PR TITLE
[index/build] Upstream indexing while building changes

### PR DESCRIPTION
**Pull request by bitjammer**
This patch upstreams previously AppleInternal changes for
indexing while building.

### CODE REVIEW

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

File: include/swift/Driver/Types.def [hunk 1]
```diff
@@ -61,6 +61,10 @@ TYPE("imported-modules", ImportedModules,   "importedmodules", "")
 TYPE("tbd",             TBD,                "tbd",             "")
 TYPE("module-trace",    ModuleTrace,        "trace.json",      "")
 
+// BEGIN APPLE-ONLY OUTPUT TYPES
```
Comment (akyrtzi) [inline include/swift/Driver/Types.def]: I think the `APPLE-ONLY` comments can be removed.

File: lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -1153,6 +1153,17 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
       OI.CompilerMode = OutputInfo::Mode::SingleCompile;
       break;
 
+    // BEGIN APPLE-ONLY OUTPUT ACTIONS
```
Comment (akyrtzi) [inline lib/Driver/Driver.cpp]: Ditto

File: lib/Driver/ToolChains.cpp [hunk 3]
```diff
@@ -231,6 +231,13 @@ ToolChain::constructInvocation(const CompileJobAction &job,
     case types::TY_TBD:
       FrontendModeOption = "-emit-tbd";
       break;
+
+    // BEGIN APPLE-ONLY OUTPUT TYPES
```
Comment (akyrtzi) [inline lib/Driver/ToolChains.cpp]: And here.

**Review by akyrtzi — COMMENTED** [review]
Other than the `APPLE-ONLY` comments, LGTM.

**Pull request merged**